### PR TITLE
feat: add sleep timer to audio player

### LINKED ISSUE #96: feat: Add sleep timer to audio player
## Context

Follow-up to #93 (in-app audio player). Sleep timer is a standard podcast app feature that automatically pauses playback after a set duration. Useful for listening in bed — set a 30-minute timer, fall asleep, and the podcast stops on its own.

**Blocked by:** #93 (persistent in-app audio player)

### Setting a Timer
- A **timer button** (e.g. `Moon` or `Timer` icon) on the player bar opens a picker with preset options:
  - 15 minutes
  - 30 minutes
  - 45 minutes
  - 60 minutes
  - End of episode
- Selecting an option starts the countdown immediately
- The timer icon changes appearance (e.g. filled, or shows remaining time) to indicate an active timer

### Active Timer
- A small countdown indicator shows remaining time on the player bar (e.g. "23:45 left" or a subtle timer badge)
- When the timer expires, audio fades out over 3 seconds and pauses (not a hard stop)
- A toast confirms: "Sleep timer — playback paused"
- The timer resets after firing (no persistent timer across sessions)

### Canceling
- Tapping the timer button while a timer is active shows "Cancel Timer" as the first option
- Selecting a new duration replaces the current timer

## Acceptance Criteria

- [ ] `VERIFY:` A sleep timer button is visible on the player bar
- [ ] `VERIFY:` Tapping it shows preset options: 15m, 30m, 45m, 60m, End of episode
- [ ] `VERIFY:` Selecting an option starts a visible countdown on the player bar
- [ ] `VERIFY:` When the timer expires, audio fades out (volume decrease over ~3 seconds) and pauses
- [ ] `VERIFY:` A toast confirms playback was paused by the sleep timer
- [ ] `VERIFY:` "End of episode" pauses at the natural end of the current episode (does not auto-play next from queue)
- [ ] `VERIFY:` An active timer can be canceled from the timer menu
- [ ] `VERIFY:` Setting a new duration replaces any active timer
- [ ] `VERIFY:` The timer icon visually indicates when a timer is active
- [ ] `VERIFY:` All existing tests pass, lint passes, build succeeds

### Step 1: Add sleep timer state to audio player context

In `src/contexts/audio-player-context.tsx`, add:
- `sleepTimer: { endTime: number; type: 'duration' | 'end-of-episode' } | null`
- `setSleepTimer(minutes: number | 'end-of-episode')` — sets the timer
- `cancelSleepTimer()` — clears the timer
- A `setInterval` (every second) that checks if `Date.now() >= endTime` and triggers the fade-out
- For "end of episode": listen for the `ended` event and pause instead of auto-playing next

### Step 2: Implement audio fade-out

When the timer fires:
1. Gradually reduce `audio.volume` from current level to 0 over 3 seconds (use `requestAnimationFrame` or small interval)
2. Once volume reaches 0, call `audio.pause()`
3. Reset volume to the user's original level (so next play isn't silent)
4. Show a toast and clear the timer state

### Step 3: Create sleep timer menu component

Create `src/components/audio-player/sleep-timer-menu.tsx`:
- A popover triggered by a timer icon on the player bar
- Shows preset duration buttons (15m, 30m, 45m, 60m, End of episode)
- When a timer is active, shows remaining time and "Cancel Timer" option
- Uses Radix `Popover` (already available via shadcn/ui)

### Step 4: Add timer button to player bar

In `src/components/audio-player/player-bar.tsx`:
- Add the sleep timer icon button (e.g. `Moon` from lucide-react)
- Show an active indicator when a timer is set (different icon color or a small badge with remaining time)
- On mobile: include in the compact controls or behind a "more" menu

## Files to Modify

| File | Change |
|------|--------|
| `src/contexts/audio-player-context.tsx` | Add sleep timer state, countdown logic, fade-out on expiry |
| New: `src/components/audio-player/sleep-timer-menu.tsx` | Timer preset picker + cancel |
| `src/components/audio-player/player-bar.tsx` | Add timer icon button with active indicator |

### In Scope
- Preset sleep timer durations (15m, 30m, 45m, 60m, end of episode)
- Visible countdown on player bar
- Graceful audio fade-out on expiry
- Cancel / replace active timer

### Out of Scope
- Custom duration input (freeform minutes)
- "Shake to extend" or motion-based timer reset
- Recurring / default timer across sessions

## Dependencies

Blocked by: #93

### PR DESCRIPTION
**Pull request by funny-otter**
## Summary

- Add sleep timer feature with preset durations (15/30/45/60 min) and "End of episode" option
- Audio fades out over 3 seconds when timer expires, then pauses playback with toast confirmation
- New `SleepTimerMenu` dropdown component with Moon icon, countdown display, and cancel option
- Timer state integrated into `AudioPlayerContext` with `Date.now()` anchoring for background tab reliability
- "End of episode" mode intercepts `onEnded` to prevent auto-play-next from queue

## Test plan

- [x] 952 tests pass (0 failures), lint clean, build succeeds
- [x] Unit tests for `fadeOutAudio` utility (7 tests): completion, cleanup, volume restore, edge cases
- [x] Unit tests for `SleepTimerMenu` component (16 tests): rendering, interactions, active states
- [x] Player bar test for sleep timer button presence
- [x] Storybook stories: Default, Active Duration Timer, Active End-of-Episode
- [x] Storybook build passes
- [ ] Manual: verify countdown display updates in real-time
- [ ] Manual: verify fade-out and pause on timer expiry
- [ ] Manual: verify "End of episode" blocks auto-play-next
- [ ] Manual: verify cancel mid-countdown and mid-fade
- [ ] Manual: verify mobile layout fits on small screens

Closes #96

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Sleep timer added to the audio player with presets (15/30/45/60 minutes) and an "End of episode" option; Moon control shows active status and appears in player controls.

* **Behavior**
  * Duration timers show remaining time; expiry performs a gentle fade-out then pauses playback; timers can be canceled.

* **Tests**
  * Comprehensive tests covering sleep-timer UI flows and fade-out behavior added.

### CODE REVIEW

File: src/lib/__tests__/audio-fade.test.ts [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+    vi.advanceTimersByTime(300);
+    expect(onComplete).toHaveBeenCalledOnce();
+    expect(audio.pause).toHaveBeenCalledOnce();
+  });
+
+  it("volume never goes below zero", () => {
+    const audio = createMockAudio(0.5);
+    const onComplete = vi.fn();
+    const originalVolumeSetter = Object.getOwnPropertyDescriptor(audio, "volume")?.set;
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/lib/__tests__/audio-fade.test.ts]: Fixed: removed unused `originalVolumeSetter` variable

File: src/lib/audio-fade.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      completed = true;
+      audio.volume = 0;
+      audio.pause();
+      audio.volume = originalVolume;
+      onComplete();
+      return;
+    }
+
+    audio.volume = Math.max(0, originalVolume - volumeStep * currentStep);
+  }, FADE_STEP_MS);
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/lib/audio-fade.ts]: Fixed in PR #174: removed all semicolons from `audio-fade.ts` to match repo style
